server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexer.java

```diff
@@ -0,0 +1,135 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.component.index;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.Uninterruptibles;
+import java.util.Arrays;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.elasticsearch.action.index.IndexRequest;
+import org.sonar.api.Startable;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.es.BulkIndexer;
+import org.sonar.server.es.EsClient;
+
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_PROJECT_UUID;
+import static org.sonar.server.component.index.ComponentIndexDefinition.INDEX_COMPONENTS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.TYPE_COMPONENT;
+
+public class ComponentIndexer implements Startable {
+
+  private final ThreadPoolExecutor executor;
+  private final DbClient dbClient;
+  private final EsClient esClient;
+
+  public ComponentIndexer(DbClient dbClient, EsClient esClient) {
+    this.executor = new ThreadPoolExecutor(0, 1, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+    this.dbClient = dbClient;
+    this.esClient = esClient;
+  }
+
+  /**
+   * Copy all components of all projects to the elastic search index.
+   * <p>
+   * <b>Warning</b>: This should only be called on an empty index. It does not purge anything.
+   */
+  public void index() {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      dbClient.componentDao()
+        .selectProjects(dbSession)
+        .stream()
+        .forEach(this::index);
+    }
+  }
+
+  /**
+   * Update the elastic search for one specific project. The current data from the database is used.
+   */
+  public void index(String projectUuid) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      purge(projectUuid);
+      index(
+        dbClient
+          .componentDao()
+          .selectByProjectUuid(projectUuid, dbSession)
+          .toArray(new ComponentDto[0]));
+    }
+  }
+
+  private void purge(String projectUuid) {
+    BulkIndexer.delete(esClient, INDEX_COMPONENTS, esClient
+      .prepareSearch(INDEX_COMPONENTS)
+      .setTypes(TYPE_COMPONENT)
+      .setFetchSource(false)
+      .setQuery(termQuery(FIELD_PROJECT_UUID, projectUuid)));
```

Comment:
This has been changed to `esClient.prepareDelete(INDEX_COMPONENTS, TYPE_COMPONENT, projectUuid)...`